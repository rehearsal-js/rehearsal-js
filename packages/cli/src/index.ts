import { URL } from 'node:url';
import { resolve } from 'node:path';
import assert from 'node:assert';
import fs from 'node:fs/promises';
import { Command } from 'commander';
import { migrateCommand } from './commands/migrate/index.js';
import { upgradeCommand } from './commands/upgrade.js';
import type { PackageJson } from 'type-fest';

const __dirname = new URL('.', import.meta.url).pathname;
const { version } = JSON.parse(
  await fs.readFile(resolve(__dirname, '../package.json'), 'utf-8')
) as PackageJson;

const program = new Command();

assert(version, 'Has a rehearsal version');

program.name('rehearsal').version(version).addCommand(migrateCommand).addCommand(upgradeCommand);

export { program as rehearsal, migrateCommand, upgradeCommand };
