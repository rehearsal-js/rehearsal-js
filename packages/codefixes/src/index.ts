export { codefixes, glintCodeFixes, applyCodeFix, ContentDelegate } from './codefixes.js';
export { isInstallPackageCommand } from './typescript-codefix-collection.js';
export { hints } from './hints.js';
export { getDiagnosticOrder } from './get-diagnostics.js';
export * from './types.js';
