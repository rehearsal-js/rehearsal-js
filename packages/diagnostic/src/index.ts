export { default as diagnose } from './diagnose';
