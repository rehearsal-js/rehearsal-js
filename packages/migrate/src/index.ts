export { default as migrate } from './migrate';
