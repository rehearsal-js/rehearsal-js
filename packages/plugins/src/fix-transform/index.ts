import ts from 'typescript';
import { type RehearsalService } from '@rehearsal/service';
import { getCodemodResult, getCommentsOnlyResult } from './utils';
import { type FixResult, type FixedFile, FixTransform as FixTransformI } from './interfaces';


export { getCodemodResult, getCommentsOnlyResult };
export { FixResult, FixedFile };

export class FixTransform implements FixTransformI {
  /** Engineer friendly message describes steps needs to be done to fix the related issue */
  hint?: string;

  /** Function to fix the diagnostic issue */
  fix?: (diagnostic: ts.DiagnosticWithLocation, service: RehearsalService) => FixResult;

  run(diagnostic: ts.DiagnosticWithLocation, service: RehearsalService): FixResult {
    return this.fix ? this.fix(diagnostic, service) : getCommentsOnlyResult(diagnostic);
  }
}

