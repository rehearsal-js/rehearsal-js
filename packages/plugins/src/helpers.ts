import { dirname, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
import Module from 'node:module';
import ts, { type FormatCodeSettings, type SourceFile } from 'typescript';
import { Options as PrettierOptions } from 'prettier';
import type { Location } from '@rehearsal/reporter';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
const require = Module.createRequire(import.meta.url);

const { SemicolonPreference, getDefaultFormatCodeSettings } = ts;

const INDEX_BUMP = 1; //bump line and column numbers from 0 to 1 for sarif reader

export function getLocation(sourceFile: SourceFile, start: number, length: number): Location {
  const { line: startLine, character: startColumn } =
    sourceFile.getLineAndCharacterOfPosition(start);
  const { line: endLine, character: endColumn } = sourceFile.getLineAndCharacterOfPosition(
    start + length
  );

  return {
    startLine: startLine + INDEX_BUMP,
    startColumn: startColumn + INDEX_BUMP,
    endLine: endLine + INDEX_BUMP,
    endColumn: endColumn + INDEX_BUMP,
  };
}

export function setProcessTTYto(setting: boolean): void {
  if (typeof process !== 'undefined') {
    process.stdout.isTTY = setting;
  }
}

export async function getFormatCodeSettingsForFile(filePath: string): Promise<FormatCodeSettings> {
  let prettierConfig: PrettierOptions | null = null;

  try {
    const prettier = importPrettier(filePath);
    prettierConfig = await prettier.resolveConfig(filePath, {
      editorconfig: true,
    });
  } catch (e) {
    // swallow the error. Prettier is not installed
  }

  const tsFormatCodeOptions = getDefaultFormatCodeSettings();

  let useSemicolons = true;
  let indentSize = tsFormatCodeOptions.tabSize ?? 2;
  let convertTabsToSpaces = true;

  if (prettierConfig) {
    useSemicolons = prettierConfig.semi !== false;
    indentSize = prettierConfig.tabWidth ?? indentSize;
    convertTabsToSpaces = prettierConfig.useTabs !== true;
  }

  return {
    ...tsFormatCodeOptions,
    baseIndentSize: indentSize,
    convertTabsToSpaces,
    indentSize,
    semicolons: useSemicolons ? SemicolonPreference.Insert : SemicolonPreference.Remove,
  };
}

function importPrettier(fromPath: string): typeof import('prettier') {
  const pkg = getPackageInfo('prettier', fromPath);
  const main = resolve(pkg.path);
  return require(main) as typeof import('prettier');
}

function getPackageInfo(packageName: string, fromPath: string): { path: string } {
  const paths = [__dirname];

  paths.unshift(fromPath);

  const packageJSONPath = require.resolve(`${packageName}/package.json`, {
    paths,
  });

  return {
    path: dirname(packageJSONPath),
  };
}
