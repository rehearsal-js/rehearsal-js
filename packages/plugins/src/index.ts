export * from './plugins';
export * from './data';
