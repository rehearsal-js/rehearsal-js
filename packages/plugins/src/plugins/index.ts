export { DiagnosticFixPlugin, DiagnosticFixPluginOptions } from './diagnostic-fix.plugin.js';
export {
  DiagnosticReportPlugin,
  DiagnosticReportPluginOptions,
} from './diagnostic-report.plugin.js';
export { GlintFixPlugin } from './glint-fix.plugin.js';
export { GlintReportPlugin } from './glint-report.plugin.js';
export { LintPlugin, LintPluginOptions } from './lint.plugin.js';
export { PrettierPlugin, PrettierPluginOptions } from './prettier.plugin.js';
export { ReRehearsePlugin, ReRehearsePluginOptions } from './re-rehearse.plugin.js';
export { ServiceInjectionsTransformPlugin } from './transform.plugin.js';
