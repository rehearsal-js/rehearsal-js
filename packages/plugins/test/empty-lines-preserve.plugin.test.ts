import { RehearsalService } from '@rehearsal/service';
import { Project } from 'fixturify-project';
import { resolve } from 'path';
import { afterEach, beforeEach, describe, expect, test } from 'vitest';

import { EmptyLinesPreservePlugin } from '../src';

describe('Test EmptyLinesPreservePlugin', function () {
  let project: Project;

  beforeEach(() => {
    project = new Project('foo', '0.0.0');
    delete project.files['index.js'];
  });

  afterEach(() => {
    project.dispose();
  });

  test('run', async () => {
    project.files['index.ts'] = `
// This is a single line comment

// This is another single line comment
const someTrueVariable = true;

/*
* This is multiline comment
*

  With some empty lines it it
*/
const someFalseVariable = false;



/**
 * This is a JSDoc comment
 *

* With some empty lines it it
*/
function someFunction(): void {

  return ;
}
    `;

    project.write();

    const files = Object.keys(project.files).map((file) => resolve(project.baseDir, file));
    const service = new RehearsalService({ baseUrl: project.baseDir }, files);

    const plugin = new EmptyLinesPreservePlugin(service);

    for (const fileName of files) {
      const result = await plugin.run({ fileName });
      const resultText = service.getFileText(fileName).trim();

      expect(result).toHaveLength(1);
      expect(resultText).toMatchSnapshot();
    }
  });
});
