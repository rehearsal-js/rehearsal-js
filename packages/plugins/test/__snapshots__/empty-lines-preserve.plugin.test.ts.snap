// Vitest Snapshot v1

exports[`Test EmptyLinesPreservePlugin > run 1`] = `
"//:line:
// This is a single line comment
//:line:
// This is another single line comment
const someTrueVariable = true;
//:line:
/*
* This is multiline comment
*
//:line:
  With some empty lines it it
*/
const someFalseVariable = false;
//:line:
//:line:
//:line:
/**
 * This is a JSDoc comment
 *
//:line:
* With some empty lines it it
*/
function someFunction(): void {
//:line:
  return ;
}
    //:line:"
`;
